diff --git a/coordinator b/coordinator
index 9dedba1..50f26c2 100755
Binary files a/coordinator and b/coordinator differ
diff --git a/coordinator.c b/coordinator.c
index 96ce0c3..6cfa61f 100644
--- a/coordinator.c
+++ b/coordinator.c
@@ -21,13 +21,64 @@ direttore di orchestra.
 #include <time.h> //per timestamp
 
 #include <pthread.h>
-
+#include <signal.h>
+#include <string.h>
+#include <sys/stat.h>
 /*
 pthread_mutex_t: tipo di dato, mutex che permette a un solo thread alla volta di accedere in una sezione critica.
 PTHREAD_MUTEX_INITIALIZER: macro, inizializza un mutex con valori di default.
 */
 pthread_mutex_t log_mutex = PTHREAD_MUTEX_INITIALIZER; //chiave per lock.
 
+#define max_Byte 1000
+
+void singalarm(int signo){
+    pthread_mutex_lock(&log_mutex);
+
+    struct stat info_file;
+
+    if(stat(LOG_FILE, &info_file) == 0){
+        if(info_file.st_size >= max_Byte){
+            printf("%ld su %d\n", info_file.st_size, max_Byte);
+            char buffer[100];
+            time_t now = time(NULL);
+            strftime(buffer, sizeof(buffer), "%d-%m-%Y_%H-%M-%S", localtime(&now));
+            char nuovo_nome[150];
+            snprintf(nuovo_nome, sizeof(nuovo_nome), "./logs/log_%s.txt", buffer);
+            rename(LOG_FILE, nuovo_nome);
+        }
+        else{
+            printf("%ld su %d\n", info_file.st_size, max_Byte);
+        }
+    }
+    else{
+        printf("errore lettura byte file\n");
+    }
+
+    pthread_mutex_unlock(&log_mutex);
+    alarm(5);
+}
+
+void write_log_dis(int id_mittente){
+    //lock
+    pthread_mutex_lock(&log_mutex);
+
+    //crea timestamp e scrivi nel log
+        FILE *f = fopen(LOG_FILE, "a");
+        if (f){
+            time_t now = time(NULL); //timestamp unix corrente
+            char ts_buffer[64];
+            strftime(ts_buffer, sizeof(ts_buffer), "%d-%m-%Y %H:%M:%S", localtime(&now)); //formatta secondo "%d-%m-%Y %H:%M:%S"
+            fprintf(f, "[%s, %d, %s]\n", ts_buffer,id_mittente, "DISCONNECT");
+            fclose(f);
+        }
+        else{
+            perror("server fopen log fallito");
+        }
+    //unlock
+    pthread_mutex_unlock(&log_mutex);
+}
+
 void write_log(int id_mittente, double dato){
     //lock
     pthread_mutex_lock(&log_mutex);
@@ -57,6 +108,15 @@ void* handle_client(void* arg){
     ssize_t n = recv(client_fd, &m, sizeof(m), 0);
     if (n == sizeof(m)){
         write_log(m.id_mittente, m.dato);
+        char *ok="ok";
+        ssize_t s = send(client_fd,ok,strlen(ok)+1,0);
+        if(s>0){
+            printf("messaggio mandato");
+        }else{
+            printf("sigpipe\n");
+            write_log_dis(m.id_mittente);
+        }
+
     }
     else if (n == 0) {  //client disconnesso
         printf("client fd = %d disconnesso senza inviare dati\n", client_fd);
@@ -73,6 +133,8 @@ void* handle_client(void* arg){
 
 int main(void){
     //creazione socket lato server, sempre in ascolto
+    signal(SIGPIPE, SIG_IGN);
+    signal(SIGALRM, singalarm);
     int listen_fd = socket(AF_INET, SOCK_STREAM, 0); //AF_INET: usa ipv4, SOCK_STREAM: usa tcp, 0: protocollo scelto automaticamente
     if (listen_fd < 0) {
         perror("socket() fallita");
@@ -102,7 +164,7 @@ int main(void){
     }
     printf("coordinator in ascolto sulla porta %d... \n", PORT);
 
-
+    alarm(5);
 
     while (1){ //loop per continuare ad accettare connessioni
         //accept(), accetta connessioni arrivate al socket.
@@ -138,7 +200,7 @@ int main(void){
 
     }
 
-
+    
     close(listen_fd);
     return 0;
 }
diff --git a/logs/log.txt b/logs/log.txt
index 999cdcf..a61220c 100644
--- a/logs/log.txt
+++ b/logs/log.txt
@@ -1,44 +1,2 @@
-[08-08-2026 21:05:26, 1, 42.50]
-[08-08-2026 21:25:37, 5, 67.00]
-[08-08-2026 21:36:22, 1, 11.00]
-[08-08-2026 21:36:38, 2, 77.00]
-[08-08-2026 22:20:07, 1, 10.00]
-[08-08-2026 22:20:07, 2, 20.00]
-[08-08-2026 22:20:07, 3, 30.00]
-[08-08-2026 22:20:07, 4, 40.00]
-[08-08-2026 22:20:07, 5, 50.00]
-[08-08-2026 22:20:07, 0, 0.00]
-[08-08-2026 22:23:24, 4, 40.00]
-[08-08-2026 22:23:24, 2, 20.00]
-[08-08-2026 22:23:24, 1, 10.00]
-[08-08-2026 22:23:24, 3, 30.00]
-[08-08-2026 22:23:24, 5, 50.00]
-[08-08-2026 22:24:31, 2, 20.00]
-[08-08-2026 22:24:31, 1, 10.00]
-[08-08-2026 22:24:31, 3, 30.00]
-[08-08-2026 22:24:31, 4, 40.00]
-[08-08-2026 22:24:31, 5, 50.00]
-[08-08-2026 22:26:53, 2, 20.00]
-[08-08-2026 22:26:53, 3, 30.00]
-[08-08-2026 22:26:53, 1, 10.00]
-[08-08-2026 22:26:53, 4, 40.00]
-[08-08-2026 22:26:53, 5, 50.00]
-[08-08-2026 22:40:08, 67, 888.00]
-[08-08-2026 22:59:41, 1, 1.00]
-[08-08-2026 22:59:56, 1, 10.00]
-[08-08-2026 22:59:56, 2, 20.00]
-[08-08-2026 22:59:56, 3, 30.00]
-[08-08-2026 22:59:56, 4, 40.00]
-[08-08-2026 22:59:56, 5, 50.00]
-[08-08-2026 23:01:29, 1, 10.00]
-[08-08-2026 23:01:29, 2, 20.00]
-[08-08-2026 23:02:32, 3, 30.00]
-[08-08-2026 23:02:32, 2, 20.00]
-[08-08-2026 23:02:32, 1, 10.00]
-[08-08-2026 23:02:32, 4, 40.00]
-[08-08-2026 23:02:32, 5, 50.00]
-[08-08-2026 23:03:35, 2, 20.00]
-[08-08-2026 23:03:35, 1, 10.00]
-[08-08-2026 23:03:35, 4, 40.00]
-[08-08-2026 23:03:35, 3, 30.00]
-[08-08-2026 23:03:35, 5, 50.00]
+[12-08-2026 22:42:14, 1, 22.00]
+[12-08-2026 22:42:15, 1, 22.00]
diff --git a/producer b/producer
index cc39a74..47b43ed 100755
Binary files a/producer and b/producer differ
